### PR TITLE
Remove default curve concept

### PR DESCRIPTION
**Pull request by richsalz**
Many tests fail, need to be looked at.
For master and 1.1.0

### CODE REVIEW

**Comment by richsalz** [comment]
WIP removed.  If client doesn't send any curves, we now treat the list as empty and don't assume that they can handle ALL curves that we can handle.  We used to assume P256, then we assuemd X25519; the safe thing is to assume nothing, although all behaviors are correct per the RFC's and all major browsers send the curves list.

**Comment by richsalz** [comment]
@davidben, your thoughts (esp around interop)

**Comment by davidben** [comment]
+1

Yay less code! We've been doing this in BoringSSL for almost a year now and I don't believe we've had any reports of breakage. This also aligns better with TLS 1.3 and is generally saner. I don't have numbers of how many clients do this, but I also don't know of an actual implementation which does it either.

You might be able test this with MITMProxy? I forget its cipher suite constraints. Offer an ECDHE_RSA and RSA version of some cipher. Test that ECDHE_RSA is chosen with a curve list and RSA is chosen without.

**Comment by vdukhovni** [comment]
What's holding this back? Another team review?

**Comment by richsalz** [comment]
Yes.  And arguably a testcase which I haven't had the time to figure out how to do.

**Comment by levitte** [comment]
Also, there are test failures in a couple of Travis builds.

**Comment by richsalz** [comment]
just pushed a rebase, let's see if that fixes things.

**Comment by vdukhovni** [comment]
Some travis builds are failing.  Perhaps because we're testing sslv3 and expecting to use ECDHE or ECDSA, but sslv3 sends no supported curves extension. The failing builds are the ones that enable ssl3. Please resolve the failing tests.  Possibly by fixing the test...

**Comment by vdukhovni** [comment]
The patch looks good FWIW, we now need to adjust the tests to match the new (more correct) behaviour.

Mind you, we must not fail to use an ECDSA certificate (supposing it is the only one we have) even if we share no curves with the client.  That's a more complex fix, since it will likely require a second pass over the shared ciphersuites with a less stringent filter. I'd suggest a separate PR for that.  This is also needed for supported signature algorithms (SHA1 cert when client only offers SHA256, ...)

**Comment by richsalz** [comment]
I added a commit to skip the EDH/ECDH tests for SSLv3.  It's a bit of a hack, but that nested map/grep thing is beyond my ken.  The root cause might be that 'openssl ciphers' (and therefore the internal cipher table) is wrong.

Your other issue, @vdukhovni I will copy into a new issue.

File: test/recipes/80-test_ssl_old.t [hunk 1]
```diff
@@ -475,8 +475,13 @@ sub testssl {
                 $flag = "-tls1_2";
             }
             foreach my $cipher (@{$ciphersuites{$protocol}}) {
-                ok(run(test([@ssltest, @exkeys, "-cipher", $cipher, $flag])),
-                "Testing $cipher");
+                if ($protocol eq "SSLv3" && $cipher =~ /EC?DH/ ) {
```
Comment (vdukhovni) [inline test/recipes/80-test_ssl_old.t]: We probably should not disable any EDH tests, just ECDH should probably suffice.  So I would drop the "?".

Comment (richsalz) [inline test/recipes/80-test_ssl_old.t]: you're right.  updated commit pushed.

**Comment by richsalz** [comment]
Done for master and 1.1.0; thanks @vdukhovni for squeezing in the review!

**Pull request closed without merging**